loader/windows/icd_windows.c

```diff
@@ -23,8 +23,75 @@
 #include <windows.h>
 #include <winreg.h>
 
+#include <initguid.h>
+#include <dxgi.h>
+typedef HRESULT (WINAPI *PFN_CREATE_DXGI_FACTORY)(REFIID, void **);
+
 static INIT_ONCE initialized = INIT_ONCE_STATIC_INIT;
 
+typedef struct WinAdapter
+{
+    char * szName;
+    LUID luid;
+} WinAdapter;
+
+const LUID ZeroLuid = { 0, 0 };
+
+static WinAdapter* pWinAdapterBegin = NULL;
+static WinAdapter* pWinAdapterEnd = NULL;
+static WinAdapter* pWinAdapterCapacity = NULL;
+
+BOOL adapterAdd(const char* szName, LUID luid)
+{
+    BOOL result = TRUE; 
+    if (pWinAdapterEnd == pWinAdapterCapacity)
+    {
+        size_t oldCapacity = pWinAdapterCapacity - pWinAdapterBegin;
+        size_t newCapacity = oldCapacity;
+        if (0 == newCapacity)
+        {
+            newCapacity = 1;
+        }
+	else 
+            newCapacity *= 2;
+
+        WinAdapter* pNewBegin = malloc(newCapacity * sizeof(*pWinAdapterBegin));
+        if (!pNewBegin)
+	    result = FALSE;
+	else
+        {
+            if (pWinAdapterBegin)
+            {
+                memcpy(pNewBegin, pWinAdapterBegin, oldCapacity * sizeof(*pWinAdapterBegin));
+                free(pWinAdapterBegin);
+            }
+            pWinAdapterCapacity = pNewBegin + newCapacity;
+            pWinAdapterEnd = pNewBegin + oldCapacity;
+            pWinAdapterBegin = pNewBegin;
+        }
+    }
+    if (pWinAdapterEnd != pWinAdapterCapacity)
+    {
+        size_t nameLen = (strlen(szName) + 1)*sizeof(szName[0]);
+        pWinAdapterEnd->szName = malloc(nameLen);
+	if (!pWinAdapterEnd->szName)
+	    result = FALSE;
+	else 
+	{
+            memcpy(pWinAdapterEnd->szName, szName, nameLen);
+            pWinAdapterEnd->luid = luid;
+            ++pWinAdapterEnd;
+        }
+    }
+    return result;
+}
+
+void adapterFree(WinAdapter *pWinAdapter)
+{
+    if(pWinAdapter->szName)
```

Comment:
You don't need to check for null, free is guaranteed to work according to the C standard.

Reply:
Keeping it to be on the safer side.

Reply:
This is not safer, this is just redundant:
https://stackoverflow.com/questions/6084218/is-it-good-practice-to-free-a-null-pointer-in-c

Reply:
Yea, though checking free isnt really needed, I regularly see crashes elsewhere in the driver when not checking if a pointer exists. Though this is probably due to bad practices that are followed elsewhere and tricked down to the crash spot... so it has become a trend to check